src/chain/shared_state.rs

```diff
@@ -377,23 +429,48 @@ impl SectionProofChain {
         }
         true
     }
-}
 
-impl Debug for SectionProofChain {
-    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
-        write!(
-            formatter,
-            "SectionProofChain(len = {})",
-            self.blocks.len() + 1
-        )
+    pub fn last_public_key(&self) -> &BlsPublicKey {
+        self.blocks
+            .last()
+            .map(|block| &block.key)
+            .unwrap_or(&self.genesis_pk)
+    }
+
+    pub fn last_index(&self) -> usize {
+        self.blocks.len()
+    }
+
+    pub fn all_keys(&self) -> impl DoubleEndedIterator<Item = &BlsPublicKey> {
+        iter::once(&self.genesis_pk).chain(self.blocks.iter().map(|block| &block.key))
+    }
+
+    pub fn len(&self) -> usize {
+        self.blocks.len() + 1
+    }
+
+    pub fn slice(&self, first_index: usize, last_index: usize) -> SectionProofChain {
+        let genesis_pk = if first_index == 0 {
+            self.genesis_pk.clone()
+        } else {
+            self.blocks[first_index - 1].key.clone()
+        };
+        let blocks = if last_index == 0 {
+            vec![]
+        } else if last_index < self.blocks.len() {
+            self.blocks[first_index..last_index].to_vec()
+        } else {
+            self.blocks[first_index..].to_vec()
+        };
+        SectionProofChain { genesis_pk, blocks }
```

Comment:
in the special case of `first_index = 0, last_index = 0`, we will return a proof chain contains the gensis_pk, but with no blocks?

Reply:
Yes, I think we can simplify this, but I'd prefer to avoid doing it in this PR.

Reply:
hmm...  I think in this PR it already does this? or you think this situation will never happen?